### PR TITLE
Highlight incompatible payment gateways

### LINKED ISSUE #7010: Enabled Payments sidebar cards render raw HTML.
## Describe the bug

Currently, when you select the payment methods block, you will see a list of active payments method on the sidebar. The sidebar renders the title and description of those methods, and it renders raw HTML.

image

## To reproduce

1. To fully see this, enable WooCommerce Payments.
2. Go to checkout editor page, select the payment methods block.
3. In the sidebar, you should see raw HTML.

## Expected behavior

HTML should either be stripped or rendered.

### LINKED ISSUE #7855: Highlight incompatible payment gateways in the C&C Blocks
## Current behavior

Currently, while using our Blocks, the shopper is not warned whether one of his payment gateways is incompatible with the C&C Blocks. Without any explanation, he will only notice that his payment gateway is missing if he checks the Checkout Block page on the front end. This can cause confusion and frustration!

## Expected behavior

Based on the results from the research spike: pdFofs-mB-p2, we decided to start implementing detecting payment gateways in the C&C Blocks and showing a notice for it:

- Implement the logic for detecting payment gateways in the C&C Blocks editor pages
- Show a notice at the top on the settings sidebar with the list of incompatible gateways (The design is available here: pca54o-4Hm-p2#comment-4760)
- Hide the `Update page settings` notice while the notification for incompatible payment gateways is visible.
- Hide the general incompatibility notice in favor of these more useful contextual notices. When the user dismisses it, we can show it again if it was not dismissed.
- Capitalizing the label for the button on the page settings notification (see [image](https://user-images.githubusercontent.com/14235870/205911264-d4579bab-6c06-4f22-b2db-6bae1c3b8417.png)):

## Screenshots

<table>
<tr>
<td>Current:
<br><br>

image
</td>
<td>Expected:
<br><br>

image
</td>
</tr>
</table>

### LINKED ISSUE #7856: Highlight incompatible payment gateways within the Payment Options Block
## Current behavior

Currently, the `Payment Options` Block displays all enabled payment gateways on the sidebar (i.e., even the incompatible ones). This will create further confusion!

## Expected behavior

We should highlight the incompatible payment gateways in the `Payment Options` Block.

## Screenshots

<table>
<tr>
<td>Current:
<br><br>

image
</td>
<td>Expected:
<br><br>

image
</td>
</tr>
</table>

This issue is related to #7010.

### PR DESCRIPTION
**Pull request by tarhi-saad**
This PR detects incompatible payment gateways with the Cart & Checkout Blocks and informs the merchant using a sidebar notice and within the Payment Options Block!

In this PR, the approach we took was to get the list of installed and **enabled** payment methods from the `allSettings.globalPaymentMethods` object. We did not include the disabled payment gateways to avoid cluttering our notice.

We initially set the `incompatiblePaymentMethods` key of the payments store to contain an object keyed by the payment method id. Each entry has a value of the payment method's title.

Then, in the reducer, when the `SET_AVAILABLE_PAYMENT_METHODS` action is dispatched, we remove items from the `incompatiblePaymentMethods` object if they have appeared in this list.

Fixes #7855
Fixes #7856
Fixes #7010

#### Highlight incompatible payment gateways using a sidebar notice

<table>
<tr>
<td>Before:
<br><br>

image
</td>
<td>After:
<br><br>

image
</td>
</tr>
</table>

#### Highlight incompatible payment gateways within the `Payment Options Block`

<table>
<tr>
<td>Before:
<br><br>

image
</td>
<td>After:
<br><br>

image
</td>
</tr>
</table>

#### Automated Tests
* [ ] Changes in this PR are covered by Automated Tests.
  * [ ] Unit tests
  * [ ] E2E tests

#### User facing testing

**Highlight incompatible payment gateways using a sidebar notice**

1. Add incompatible payment gateways with the C&C Blocks (e.g., Paypal, Klarna Payments). You will need to configure them to be detected and used by WooCommerce.
2. Create a new page and add the `Checkout Block`
3. Verify there is a notice in the sidebar displaying the list of incompatible gateways. The `make as default` & `sidebar compatibility` notices should be hidden while the `incompatible gateways` notice is displayed.

image

4. Select each inner block in the Checkout Blocks' `List View` to ensure our notice is displayed for each one.
5. Dismiss the `incompatible gateways` notice by clicking the top right `X` button. The `make as default` & `sidebar compatibility` notices should be visible again.

image

6. Do the same test from step 2 to 5 using the `Cart Block`

**Highlight incompatible payment gateways within the `Payment Options Block`**

1. If it's not done already, add incompatible payment gateways with the C&C Blocks (e.g., Paypal, Klarna Payments)
2. Create a new page and add the `Checkout Block`
3. Select the `Payment Options Block`, and check the list of available payment gateways under `Methods` in the sidebar. The incompatible payment gateways should be highlighted (i.e., `Incompatible with block-based checkout`)

image

### WooCommerce Visibility

* [x] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental
 
### Changelog

> Highlight incompatible payment gateways in the Cart & Checkout Blocks

### CODE REVIEW

File: assets/js/editor-components/default-notice/index.tsx [hunk 1]
```diff
@@ -189,3 +190,45 @@ export function LegacyNotice( { block }: { block: string } ) {
 		</Notice>
 	);
 }
+
+export function IncompatibilityPaymentGatewaysNotice() {
```
Comment (opr) [inline assets/js/editor-components/default-notice/index.tsx]: I think we should rename this to `IncompatiblePaymentGatewaysNotice`

File: assets/js/editor-components/default-notice/index.tsx [hunk 2]
```diff
@@ -189,3 +190,45 @@ export function LegacyNotice( { block }: { block: string } ) {
...
+	const [ settingStatus, setStatus ] = useState( 'pristine' );
+
+	if (
+		IncompatibilityPaymentGateways.length === 0 ||
+		settingStatus === 'dismissed'
+	) {
+		return null;
+	}
+
+	const noticeContent = __(
```
Comment (opr) [inline assets/js/editor-components/default-notice/index.tsx]: We can use [`_n`](https://developer.wordpress.org/reference/functions/_n/) here to decide whether to show the plural or singular form of `gateway`. The JS function has the same API as the PHP one

File: assets/js/editor-components/default-notice/index.tsx [hunk 3]
```diff
@@ -189,3 +190,45 @@ export function LegacyNotice( { block }: { block: string } ) {
...
 	);
 }
+
+export function IncompatibilityPaymentGatewaysNotice() {
+	// Everything below works the same for Cart/Checkout
+	const { IncompatibilityPaymentGateways } = useSelect( ( select ) => {
+		const { getIncompatiblePaymentMethodData } =
+			select( PAYMENT_STORE_KEY );
+		return {
+			IncompatibilityPaymentGateways: getIncompatiblePaymentMethodData(),
```
Comment (opr) [inline assets/js/editor-components/default-notice/index.tsx]: Perhaps `incompatibleGateways` would be a better name for this?

**Comment by elizaan36** [comment]
hey @tarhi-saad Sorry to make a change at the last minute but this notice needs a slight tweak to the placement of the icon to align with the error notice system that's currently in design.
Notice Banner

Here's a screenshot of the new placement - rather than using multiple error icons for each incompatible payment method, the icon appears horizontally aligned with the top line of text and the "X" to close at the top of the notice.

**Review by nielslange — COMMENTED** [review]
I've installed and activated the following payment plugins:

- [Mollie Payments for WooCommerce](https://wordpress.org/plugins/mollie-payments-for-woocommerce/)
- [WooCommerce PayPal Payments](https://wordpress.org/plugins/woocommerce-paypal-payments/)
- [WooCommerce Stripe Gateway](https://wordpress.org/plugins/woocommerce-gateway-stripe/)

Yet, I can only see the following payment plugin within the Checkout block:

- [Mollie Payments for WooCommerce](https://wordpress.org/plugins/mollie-payments-for-woocommerce/)

Screenshot 2022-12-23 at 15 46 49

Do you have any idea what might causes this?

File: assets/js/blocks/checkout/inner-blocks/checkout-payment-block/edit.tsx [hunk 4]
```diff
@@ -9,6 +9,9 @@ import { ADMIN_URL, getSetting } from '@woocommerce/settings';
 import ExternalLinkCard from '@woocommerce/editor-components/external-link-card';
 import { innerBlockAreas } from '@woocommerce/blocks-checkout';
 import Noninteractive from '@woocommerce/base-components/noninteractive';
+import { GlobalPaymentMethod } from '@woocommerce/type-defs/payments';
```
Comment (nielslange) [inline assets/js/blocks/checkout/inner-blocks/checkout-payment-block/edit.tsx]:
```suggestion
import { GlobalPaymentMethod } from '@woocommerce/types';
```

With https://github.com/woocommerce/woocommerce-blocks/pull/8019, I created a PR to simplify the `@woocommerce/types` imports.

File: assets/js/blocks/checkout/inner-blocks/checkout-payment-block/edit.tsx [hunk 5]
```diff
@@ -38,9 +35,18 @@ export const Edit = ( {
...
+	);
+
+	const { incompatiblePaymentMethods } = useSelect( ( select ) => {
+		const { getIncompatiblePaymentMethods } = select( PAYMENT_STORE_KEY );
+		return {
+			incompatiblePaymentMethods: getIncompatiblePaymentMethods(),
+		};
+	}, [] );
+	const incompatiblePaymentMethodMessage =
+		'Incompatible with block-based checkout';
```
Comment (nielslange) [inline assets/js/blocks/checkout/inner-blocks/checkout-payment-block/edit.tsx]:
```suggestion
		__( 'Incompatible with block-based checkout', 'woo-gutenberg-products-block' );
```

I blieve that all messages should be translatable by default.

File: assets/js/data/payment/default-state.ts [hunk 6]
```diff
@@ -40,6 +55,7 @@ export const defaultPaymentState: PaymentState = {
 		Record< string, SavedPaymentMethod[] > | EmptyObjectType
 	>( 'customerPaymentMethods', {} ),
 	paymentMethodData: {},
+	incompatiblePaymentMethods,
```
Comment (nielslange) [inline assets/js/data/payment/default-state.ts:58]: Don't we have to initialize attribute? I see that all other attributes have a default value, that's why I wonder why this one is empty.

File: assets/js/editor-components/external-link-card/index.tsx [hunk 7]
```diff
@@ -15,6 +16,7 @@ export interface ExternalLinkCardProps {
 	href: string;
 	title: string;
 	description?: string;
+	warning?: string;
```
Comment (nielslange) [inline assets/js/editor-components/external-link-card/index.tsx:19]: Can this value be `undefined`? If so, we should change it to `warning?: string | undefined;` based on https://github.com/woocommerce/woocommerce-blocks/blob/trunk/tsconfig.base.json#L19

**Comment by tarhi-saad** [comment]
Thank you, @nielslange, for the review and test! Did you set up an account for Klarna and Paypal (you can enable "test mode" by using a test account)?
 
- For Klarna, you can find some instructions here: [Configuring Klarna Checkout on WooCommerce](https://docs.klarna.com/platform-solutions/woocommerce/klarna-checkout/configuring-klarna-checkout-on-woocommerce/)
- And Paypal, [Account Setup and Onboarding](https://woocommerce.com/document/woocommerce-paypal-payments/#account-setup)

**Review by nielslange — CHANGES_REQUESTED** [review]
Thanks for working on this, @tarhi-saad! 👏

I left a few comments below. The only suggestion, I'd consider serious, is the string that cannot be translated currently. I believe we should ensure that all strings are translatable.

Besides that, I also noticed that in your screenshot, the `make as default` & `sidebar compatibility` notices are shown on top of the sidebar. On my end, they show at the bottom. Especially on when viewing the `Payment Options`, it's easy to miss these two notices:

7412-notices

**Comment by nielslange** [comment]
> Thank you, @nielslange, for the review and test! Did you set up an account for Klarna and Paypal (you can enable "test mode" by using a test account)?
> 
> * For Klarna, you can find some instructions here: [Configuring Klarna Checkout on WooCommerce](https://docs.klarna.com/platform-solutions/woocommerce/klarna-checkout/configuring-klarna-checkout-on-woocommerce/)
> * And Paypal, [Account Setup and Onboarding](https://woocommerce.com/document/woocommerce-paypal-payments/#account-setup)

I strictly followed the test instructions, which didn't mention to set up the accounts. Maybe we should update the test instructions accordingly, so that other folks in the future know how to test this PR without running into the same issue I just ran into. That said, after activating the test mode, I'm now able to see other payment extensions as well.

Screenshot 2022-12-23 at 16 44 08

**Comment by tarhi-saad** [comment]
Thank you, @elizaan36! The design and screenshots are updated!

> I strictly followed the test instructions, which didn't mention to set up the accounts. Maybe we should update the test instructions accordingly, so that other folks in the future know how to test this PR without running into the same issue I just ran into. That said, after activating the test mode, I'm now able to see other payment extensions as well.

That's great, @nielslange! I updated the description accordingly! 🙌

File: assets/js/data/payment/default-state.ts (hunk 6 above)
Comment (tarhi-saad) [reply assets/js/data/payment/default-state.ts:58]: Hi @nielslange! The `incompatiblePaymentMethods` is initialized in line 38:

`const incompatiblePaymentMethods: Record< string, string > = {};`

File: assets/js/editor-components/external-link-card/index.tsx (hunk 7 above)
Comment (tarhi-saad) [reply assets/js/editor-components/external-link-card/index.tsx:19]: @nielslange, this type is already defined as an [optional Parameters](https://www.typescriptlang.org/docs/handbook/2/functions.html#optional-parameters). The following type declarations are identical:

- `warning?: string | undefined;`
- `warning?: string;`

**Comment by tarhi-saad** [comment]
Thank you, @nielslange, for the review! I applied the necessary changes! 🙌

**Review by wavvves — APPROVED** [review]
Nice work @tarhi-saad!

**Pull request merged**